### ISSUE TITLE
[QUESTION] Request rate limiting

### ISSUE THREAD

**Issue by Unbrick** [realsirjoe/instagram-scraper#124]
Hey guys,

i have been experimenting with this scraper for about a week now. Basically it works like a charm, but i'm a little confused about the request rates possible. My setup uses a phone verified instagram account (through a private proxy) which automatically completes the 2 factor authentication using SMS (also through the proxy, IPs should be identical).

When requesting 5000 media items of one instagram page, the requests are split internally in 12 media items per request.  Rate limiting does not seem to kick in here.

My issue pops up after requesting media (independent of the count, does not matter if 50 or 5000) of any instagram page. 
Scraping one page works like a charm, subsequent page scrapes will throw an error:

```
InstagramException: Response code is 403. Body: {"message": "login_required", "error_title": "You've Been Logged Out", "error_body": "Please log back in.", "logout_reason": 2, "status": "fail"}
```

Logging in again does work (no two factor reqired) but i cant log in for every page i scrape as logging in rate limits at 8-10 times in a short period of time. 

Has anyone a simmilar behaviour or knows what the issue here is?

**Comment by lukleu** [comment]
Same problem, no solution.. are you multithreading?

**Comment by Unbrick** [comment]
Nope, scraping single threaded. Currently 24 posts per page with `sleep_between_requests=5`

**Comment by lukleu** [comment]
Lets investigate together then?

**Comment by Unbrick** [comment]
I'll try to reproduce it in a debugger console, i'll post anything i find!

**Comment by lukleu** [comment]
Pretty sure its not the actual request limit .. im having the problem with requests with 300 seconds in between when crawling posts

**Comment by Unbrick** [comment]
I have been intercepting the http connection.

Seems like there are three different types of API to communicate to. 
- Graphql API (`/graphql/...`)
- Normal API (`/api/v1/...`)
- Web frontend

Communicating with the Graphql and normal API does not bring any problems in request limiting or similar. The error gets thrown when communicating with the web frontend. Despite of beeing logged out from the web frontend, requests to the other two APIs can still be made.

The issue we have is that any request to resolve a username to the user id works by utilizing the web frontend. Not sure what the issue exactly is there but try scraping all instagram user IDs and saving them. Then make requests using the `get_medias_by_user_id` method, this one only triggers the graphql API and should therefore work.

**Comment by lukleu** [comment]
Which methods to avoid then? 🤔 or could we rewrite?

**Comment by Unbrick** [comment]
As far as i know you should avoid every method which has to do with login (except the `with_credentials()` and the `login()` itself of course) and any method taking in a username as a parameter. Not how many these are.

As i'm running on a headless debian server i used mitmproxy in console mode. It also turned out that this api has a bug / race condition where some requests are sent with the `User-Agent:       python-requests/2.24.0` header instead of the spoofed header (Android 6.0.1 etc..) but i dont think thats the root cause of our issue here.

**Comment by lukleu** [comment]
my guess would be that it is indeed the incorrect header - instagram might interpret this as the session changing from one device to another without logging back in

**Comment by Unbrick** [comment]
I just wanted to report back my currently working setup. I'm using a phone verified account using a online phone number from twilio (fully automated 2fa ftw!). Also i'm using a private proxy for scraping but I'm not changing the proxy while scraping so seems like i'm not triggering the IP rate limiting. Also i use the `sleep_between_requests = 2` to space out the requests a little.

I'm scraping the account identifiers (using `get_account()`) in the first stage and afterwards i'm just using the `get_medias_by_user_id()` and the `get_media_by_id()` methods. In this order scraping works fine, mixing up the method calls (calling `get_account()` after the `get_medias_by_user_id()`) does not work. Seems like the `get_medias_by_user_id()` screws up some cookies which makes the session invalid.

**Comment by lukleu** [comment]
I'm investigating a similar train of thought, I've split up my queues between threads and it seems that certain functions are more likely to break the session - since I'm using multiple accounts and proxies it's not as easy to identify though

**Comment by Parham-sagharchi** [comment]
> I'm investigating a similar train of thought, I've split up my queues between threads and it seems that certain functions are more likely to break the session - since I'm using multiple accounts and proxies it's not as easy to identify though

could u explain more ?

**Comment by schaumju** [comment]
> I'm investigating a similar train of thought, I've split up my queues between threads and it seems that certain functions are more likely to break the session - since I'm using multiple accounts and proxies it's not as easy to identify though
@lukasleuschen @Unbrick 
Where you by any chance able to figure this out? I'm trying to get the followers of large accounts (like millions) so splitting it up into threads to get the followers seems like the only way to do it in a feasible amount of time.
